akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldResourceAsyncSinkSpec.scala

```diff
@@ -0,0 +1,211 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import java.io.{ BufferedWriter, File, FileWriter, Writer }
+import java.util.concurrent.Executors
+
+import akka.Done
+import akka.stream.ActorAttributes._
+import akka.stream.Supervision._
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.Utils._
+import akka.stream.testkit.scaladsl.TestSource
+import org.scalatest.BeforeAndAfter
+
+import scala.concurrent.{ Await, ExecutionContext, Future, Promise }
+
+class FoldResourceAsyncSinkSpec extends StreamSpec(UnboundedMailboxConfig) with BeforeAndAfter {
+  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
+  implicit val materializer = ActorMaterializer(settings)
+
+  val ioExecutor = Executors.newFixedThreadPool(4)
+  val ioExecutionContext = ExecutionContext.fromExecutor(ioExecutor)
+
+  override def afterTermination(): Unit = {
+    ioExecutor.shutdown()
+  }
+
+  val manyLines = {
+    ("a" * 100 + "\n") * 10 +
+      ("b" * 100 + "\n") * 10 +
+      ("c" * 100 + "\n") * 10 +
+      ("d" * 100 + "\n") * 10 +
+      ("e" * 100 + "\n") * 10 +
+      ("f" * 100 + "\n") * 10
+  }
+  val manyLinesArray = manyLines.split("\n")
+
+  var sinkFile: File = _
```

Comment:
Please avoid using filesystem from the test like this if possible since it may be slow on the ci-servers and cause unstable tests. All scenarios should be possible to emulate using the testkit probes.